File: src/main/java/org/tud/oas/population/PopulationView.java
package org.tud.oas.population;

import org.locationtech.jts.geom.*;
import org.locationtech.jts.index.kdtree.*;
import org.locationtech.jts.algorithm.locate.*;

import java.util.ArrayList;
import java.util.List;

public class PopulationView implements IPopulationView {
    private KdTree index;
    private List<Coordinate> points;
    private List<Coordinate> utm_points;
    private List<Integer> counts;

    private Geometry area;
    private Envelope envelope;

    public PopulationView(List<Coordinate> points, List<Coordinate> utm_points, List<Integer> counts,
            Envelope envelope) {
        this.points = points;
        this.utm_points = utm_points;
        this.counts = counts;
        this.index = new KdTree();
        int i = 0;
        for (Coordinate coord : points) {
            this.index.insert(coord, i);
            i++;
        }
        this.envelope = envelope;
        this.area = null;
    }

    public PopulationView(List<Coordinate> points, List<Coordinate> utm_points, List<Integer> counts, Geometry area) {
        this.points = points;
        this.utm_points = utm_points;
        this.counts = counts;
        this.index = new KdTree();
        int i = 0;
        for (Coordinate coord : points) {
            this.index.insert(coord, i);
            i++;
        }
        if (area != null) {
            this.area = area;
            this.envelope = area.getEnvelopeInternal();
        }
    }

    public Coordinate getCoordinate(int index) {
        return this.points.get(index);
    }

    public Coordinate getCoordinate(int index, String crs) {
        if ("EPSG:4326".equals(crs)) {
            return this.points.get(index);
        } else if ("EPSG:25832".equals(crs)) {
            return this.utm_points.get(index);
        }
        return new Coordinate(0, 0);
    }

    public int getPopulation(int index) {
        return this.counts.get(index);
    }

    public int pointCount() {
        return this.points.size();
    }

    public List<Integer> getPointsInEnvelop(Envelope envelope) {
        Envelope env;
        if (this.envelope == null) {
            env = envelope;
        } else {
            env = this.envelope.intersection(envelope);
        }
        if (env == null) {
            List<Integer> result = new ArrayList<>();
            for (int i = 0; i < pointCount(); i++) {
                result.add(i);
            }
            return result;
        }

        List<Integer> points = new ArrayList<>(100);

        this.index.query(env, node -> {
            if (this.area == null) {
                int index = (int) node.getData();
                points.add(index);
            } else {
                int location = SimplePointInAreaLocator.locate(node.getCoordinate(), this.area);
                if (location == Location.INTERIOR) {
                    int index = (int) node.getData();
                    points.add(index);
                }
            }
        });

        return points;
    }
}

File: src/population/PopulationView.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NetTopologySuite.Geometries;
using NetTopologySuite.Index.KdTree;
using NetTopologySuite.Algorithm.Locate;

namespace DVAN.Population
{
    public class PopulationView : IPopulationView
    {
        public KdTree<object> index;
        public List<Coordinate> points;
        public List<Coordinate> utm_points;
        public List<int> counts;

        Geometry? area;
        Envelope? envelope;

        public PopulationView(List<Coordinate> points, List<Coordinate> utm_points, List<int> counts, Envelope? envelope)
        {
            this.points = points;
            this.utm_points = utm_points;
            this.counts = counts;
            this.index = new KdTree<object>();
            int i = 0;
            foreach (var coord in points) {
                this.index.Insert(coord, i);
                i++;
            }
            this.envelope = envelope;
            this.area = null;
        }

        public PopulationView(List<Coordinate> points, List<Coordinate> utm_points, List<int> counts, Geometry? area)
        {
            this.points = points;
            this.utm_points = utm_points;
            this.counts = counts;
            this.index = new KdTree<object>();
            int i = 0;
            foreach (var coord in points) {
                this.index.Insert(coord, i);
                i++;
            }
            if (area != null) {
                this.area = area;
                this.envelope = area.EnvelopeInternal;
            }
        }

        public Coordinate getCoordinate(int index)
        {
            return this.points[index];
        }

        public Coordinate getCoordinate(int index, String crs)
        {
            if (crs == "EPSG:4326") {
                return this.points[index];
            }
            else if (crs == "EPSG:25832") {
                return this.utm_points[index];
            }
            return new Coordinate(0, 0);
        }

        public int getPopulation(int index)
        {
            return this.counts[index];
        }

        public int pointCount()
        {
            return this.points.Count;
        }

        public List<int> getPointsInEnvelop(Envelope? envelope)
        {
            Envelope env;
            if (this.envelope == null) {
                env = envelope;
            }
            else {
                env = this.envelope.Intersection(envelope);
            }
            if (env == null) {
                return Enumerable.Range(0, this.pointCount()).ToList();
            }

            List<int> points = new List<int>(100);
            var visitor = new VisitKdNode<object>();
            visitor.setFunc((node) => {
                if (this.area == null) {
                    int index = (int)node.Data;
                    points.Add(index);
                }
                else {
                    Location location = SimplePointInAreaLocator.Locate(node.Coordinate, this.area);
                    if (location == Location.Interior) {
                        int index = (int)node.Data;
                        points.Add(index);
                    }
                }
            });

            this.index.Query(env, visitor);

            return points;
        }
    }
}